diesel/src/query_dsl/mod.rs

```diff
@@ -287,6 +291,134 @@ pub trait QueryDsl: Sized {
+    /// // By default, all columns will be selected
+    /// let all_users = users.load::<(i32, String)>(&connection)?;
+    /// assert_eq!(vec![(1, String::from("Sean")), (2, String::from("Tess"))], all_users);
+    ///
+    /// let all_names = users.select_by::<UserName>().load::<UserName>(&connection)?;
+    /// assert_eq!(vec![UserName("Sean".into()), UserName("Tess".into())], all_names);
+    /// #     Ok(())
+    /// # }
+    /// ```
+    ///
+    /// ### When used with a left join
+    ///
+    /// ```rust
+    /// # include!("../doctest_setup.rs");
+    /// # use schema::{users, posts};
+    /// #
+    /// # #[derive(Queryable, PartialEq, Eq, Debug)]
+    /// # struct User {
+    /// #     id: i32,
+    /// #     name: String,
+    /// # }
+    /// #
+    /// # impl User {
+    /// #     fn new(id: i32, name: &str) -> Self {
+    /// #         User {
+    /// #             id,
+    /// #             name: name.into(),
+    /// #         }
+    /// #     }
+    /// # }
+    /// #
+    /// # #[derive(PartialEq, Eq, Debug, Clone, Queryable, QueryableByColumn)]
+    /// # #[table_name = "users"]
+    /// # pub struct UserName(#[column_name = "name"] String);
+    /// #
+    /// # #[derive(Queryable, PartialEq, Eq, Debug)]
+    /// # struct Post {
+    /// #     id: i32,
+    /// #     user_id: i32,
+    /// #     title: String,
+    /// # }
+    /// #
+    /// # impl Post {
+    /// #     fn new(id: i32, user_id: i32, title: &str) -> Self {
+    /// #         Post {
+    /// #             id,
+    /// #             user_id,
+    /// #             title: title.into(),
+    /// #         }
+    /// #     }
+    /// # }
+    /// #
+    /// # #[derive(PartialEq, Eq, Debug, Clone, Queryable, QueryableByColumn)]
+    /// # #[table_name = "posts"]
+    /// # pub struct PostTitle(#[column_name = "title"] String);
+    /// #
+    /// # fn main() {
+    /// #     run_test().unwrap();
+    /// # }
+    /// #
+    /// # fn run_test() -> QueryResult<()> {
+    /// #     let connection = establish_connection();
+    /// #     connection.execute("DELETE FROM posts")?;
+    /// #     diesel::insert_into(posts::table)
+    /// #         .values((posts::user_id.eq(1), posts::title.eq("Sean's Post")))
+    /// #         .execute(&connection)?;
+    /// #     let post_id = posts::table.select(posts::id)
+    /// #         .first::<i32>(&connection)?;
+    /// let join = users::table.left_join(posts::table);
+    ///
+    /// // By default, all columns from both tables are selected
+    /// let all_data = join.load::<(User, Option<Post>)>(&connection)?;
+    /// let expected_data = vec![
+    ///     (User::new(1, "Sean"), Some(Post::new(post_id, 1, "Sean's Post"))),
+    ///     (User::new(2, "Tess"), None),
+    /// ];
+    /// assert_eq!(expected_data, all_data);
+    ///
+    /// // Since `PostTitle` is on the right side of a left join, `Option` is
+    /// // needed.
+    /// let names_and_titles = join.select_by::<(UserName, Option<PostTitle>)>()
```

Comment:
Yes, you have to type for both if it could not be inferred by context, theoretically, you could `.select_by::<A>()` while `.load::<B>()` if and only if `A` has the same `SelectExpression` as `B`.

Reply:
I would be quite surprised if something like `.select_by::<A>().load::<B>()` works as the corresponding `LoadDsl` impl clearly states that
```rust
impl<Conn, S, Stmt> LoadQuery<Conn, S> for SelectByStatement<S, Stmt>
```
`A` (The first type parameter for `SelectByStatement`) and `B` (the second type parameter for `LoadQuery`) needs to be the same. The only question is: Is rustc's type inference able to figure that out through the layer of `RunQueryDsl` and other abstractions or does it require both annotations.

Reply:
I'm wrong, `.select_by::<A>().load::<B>()` would not work, I think compiler could infer `A` that based by my test (i.e. `let _ = users.select_by::<A>().load()` compiles).

Reply:
I mean we could make `.select_by::<A>().load::<B>()` working at the cost of requiring both type annotations in all cases, but I think the simplicity of just having one type annotation there is a good argument for not doing this.
Beside of that I think it should even work if you don't specify both type arguments as long as the return type can be inferred (for example by returning the result from a function, or by just having a type annotation on the let binding). It would probably be a good idea to have a "test case" for all those different variants in the existing test cases, to see if those work and to ensure they won't break in the future.